File: src-python/singly_linked_list.py
from typing import TypeVar, Generic


E = TypeVar('E')


class SingleLinkedList(Generic[E]):

    class _Node:
        def __init__(self, data: E | None):
            self.data = data
            self.next: SingleLinkedList._Node | None = None

    def __init__(self):
        self._dummyHead = self._Node(None)
        self._size = 0

    def add(self, index: int, e: E) -> None:
        if index < 0 or index > self._size:
            raise IndexError("索引越界")

        prev = self._dummyHead
        for _ in range(index):
            prev = prev.next

        new_node = self._Node(e)
        new_node.next = prev.next
        prev.next = new_node
        self._size += 1

    def remove(self, index: int) -> E:
        if index < 0 or index >= self._size:
            raise IndexError("索引越界")

        prev = self._dummyHead
        for _ in range(index):
            prev = prev.next

        del_node = prev.next
        prev.next = del_node.next
        del_node.next = None
        self._size -= 1
        return del_node.data

    def print_singly_linked_list(self) -> None:
        cur = self._dummyHead.next
        while cur.next is not None:
            print(cur.data, end=" -> ")
            cur = cur.next
        print(cur.data)

if __name__ == '__main__':
    list = SingleLinkedList()
    list.add(0, 0)
    list.add(1, 1)
    list.add(2, 2)
    list.add(3, 3)
    list.add(4, 4)
    list.add(5, 5)
    list.add(6, 6)
    list.add(7, 7)
    list.add(8, 8)
    list.add(9, 9)
    list.print_singly_linked_list()

File: src-java/src/main/java/SinglyLinkedList.java
/**
 * ClassName: SinglyLinkedList
 * Package: PACKAGE_NAME
 * Description:
 *
 * @Author AnXin
 * @Create 2026/5/29 15:53
 * @Version 1.0
 */
public class SinglyLinkedList<E> {

    private final Node dummyHead;
    private int size;

    public SinglyLinkedList() {
        dummyHead = new Node(null);
        size = 0;
    }

    private class Node {
        E data;
        Node next;

        Node() {

        }

        Node(E data) {
            this.data = data;
            this.next = null;
        }

        Node(E data, Node next) {
            this.data = data;
            this.next = next;
        }
    }

    public void add(int index, E e) {
        if (index < 0 || index > size) {
            throw new IndexOutOfBoundsException("索引越界");
        }

        Node prev = dummyHead;
        for (int i = 0; i < index; i++) {
            prev = prev.next;
        }

        Node newNode = new Node(e);
        newNode.next = prev.next;
        prev.next = newNode;
        size++;
    }

    public E remove(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("索引越界");
        }

        Node prev = dummyHead;
        for (int i = 0; i < index; i++) {
            prev = prev.next;
        }

        Node delNode = prev.next;
        prev.next = delNode.next;
        delNode.next = null;
        size--;
        return delNode.data;
    }

    public void printSinglyLinkedList() {
        Node cur = dummyHead.next;
        while (cur.next != null) {
            System.out.print(cur.data + " -> ");
            cur = cur.next;
        }
        System.out.println(cur.data);
    }
}
